### ISSUE TITLE
FCM times out blocking Allow Notifications UI

### ISSUE THREAD

**Issue by advayDev1** [WorldHealthOrganization/app#874]
```
W/FirebaseInstanceId( 9328): Token retrieval failed without exception message. Will retry token retrieval
W/MessengerIpcClient( 9328): Timing out request: 1
```

This keeps the user stuck on the Allow Notifications screen.
Instead, this should be done asynchronously when available.

Furthermore, the app never registers for FCM token changes, which is required to async send the token to AppEngine when it does get eventually set.

**Comment by brunobowden** [comment]
@advayDev1 - asynchronous is clearly better, though I get wary of changes like that closer to launch

**Comment by advayDev1** [comment]
Pat please take a look at this.  There's no need to hold this screen on FCM tokens.

**Comment by advayDev1** [comment]
@brunobowden i think not doing it is a bigger risk, of trapping ppl on the 2nd screen of the app.

**Comment by patniemeyer** [comment]
I think the fix to this is straightforward but I'm not in a good position to test it right now so we may have to wait for @cbgrey 's next availability.   Assuming it's blocking in `registerFCMToken()` on the `getToken()` call we should do two things: 

1) remove the `await` on line 43 where we call `registerFCMToken()`. instead allowing this first registration process to attempt to complete async.  (and comment the code as such)  

2) I think we probably need to repeat this call on every launch after the permission is already granted.  I'm not sure if the Firebase token value can change (the iOS APNS token can in theory but I'm not sure that they are related or if FB sends the APNS token separately in the background) but at least repeating will cover the case where we fail to register it due to network problems on this first launch.  

To accomplish #2 we should factor out this method and call it from `main` at an appropriate time if the permission was granted.  (We can either check the iOS permission status if the API allows us or store a flag ourselves).  @crazybob If you are concerned about updating the token on every app launch we could store a "last updated" value to limit the frequency.

**Comment by cbgrey** [comment]
Agree we should implement Pat's first suggestion. Not a huge risk for that change.

**Comment by cbgrey** [comment]
Looking into the other issue of ensuring FCM token is up to date. Docs have some detailed guidance for iOS and Android but less clear for Flutter.

**Comment by epicfaace** [comment]
@patniemeyer 

> 2\. I think we probably need to repeat this call on every launch after the permission is already granted.  I'm not sure if the Firebase token value can change (the iOS APNS token can in theory but I'm not sure that they are related or if FB sends the APNS token separately in the background) but at least repeating will cover the case where we fail to register it due to network problems on this first launch.


How about this -- we store a value in UserPreferences called "FCMTokenRegistered". After we call _registerFCMToken and WhoService.putDeviceToken successfully, FCMTokenRegistered is set to true. When FCMTokenRegistered is set to false, we will not call _registerFCMToken on app startup.

This way, once the token has been properly registered and sent to our backend, we won't call the backend again.

**Comment by advayDev1** [comment]
We also need to handle the changing of the token over time otherwise users will be stuck without notifications.  https://pub.dev/documentation/firebase_messaging/latest/firebase_messaging/FirebaseMessaging/onTokenRefresh.html

If we aren't already.

**Comment by epicfaace** [comment]
> We also need to handle the changing of the token over time otherwise users will be stuck without notifications. https://pub.dev/documentation/firebase_messaging/latest/firebase_messaging/FirebaseMessaging/onTokenRefresh.html
> 
> If we aren't already.

That's a good point. We aren't handling this currently.

How will changing of tokens affect the backend? Would it be okay just to call WhoService.putDeviceToken for the new token?

**Comment by patniemeyer** [comment]
I think this is complete, isn't it?

**Comment by epicfaace** [comment]
Right. I'm not sure if we actually tested to make sure this particular issue was fixed, though, when reviewing #914 -- although #914 will most probably resolve this.

**Comment by patniemeyer** [comment]
#914 should resolve this.  I don't know what we have a reasonable way to test it other than waiting for Firebase to update a token, which may not happen in normal use.

**Comment by hspinks** [comment]
Closed in #914
